File: Semestres/11.Febrero-Junio 2022/EstructurasDatosUSB/corte1/tema1colas/Cola.py
import numpy as np
class Cola:
    def __init__(self,size):
        self.head = 0 #Cabeza
        self.tail = 1 #Cola
        self.num_elements = 0
        self.array = np.zeros(size)
        self.size = size

    def enqueue(self,elemento):
        if self.num_elements < self.size:
            self.array[self.tail-1] = elemento
            self.num_elements += 1
            if(self.tail==self.size):
                self.tail = 1
            else:
                self.tail += 1

            if self.num_elements == 0:
                self.head = 1
        else:
            print("Overflow")

    def dequeue(self):
        if self.num_elements == 0:
            print("Underflow")
        else:
            elemento = self.array[self.head-1]
            
            if(self.head == self.size):
                self.head = 1
            else:
                self.head+=1

            self.num_elements-=1
            return elemento

File: Semestres/12. Agosto-Diciembre 2022/FADA/Corte1/5.EstructurasDatos/2.PILACOLA/Cola.java
public class Cola {
  int head;
  int tail;
  int size_max;
  int size;
  int arreglo[];

  Cola(int n){
    this.head = 1;
    this.size_max = n;
    this.size = 0;
    this.tail = 1;
    this.arreglo = new int[n]; 
  }

  public void enqueue(int n) throws IndexOutOfBoundsException{
    if (this.size_max == size){
      throw new IndexOutOfBoundsException();
    }
    else{
      this.size++;
      arreglo[this.tail-1] = n;
      this.tail = (this.tail>=this.size_max) ? 1 : this.tail+1;
    }
  }

  public int dequeue() throws IndexOutOfBoundsException{
    if (this.size == 0){
      throw new IndexOutOfBoundsException();
    }
    else{
      this.size--;
      int val = arreglo[this.head-1];
      this.head = (this.head>=this.size_max) ? 1 : this.head+1;
      return val;
      
    }
  }
}
